src/runtime_src/core/common/usage_metrics.cpp

```diff
@@ -0,0 +1,515 @@
+#include "core/common/shim/buffer_handle.h"
+#include "core/common/shim/hwctx_handle.h"
+#include "core/include/xrt/xrt_uuid.h"
+
+#include <algorithm>
+#include <boost/property_tree/json_parser.hpp>
+#include <boost/property_tree/ptree.hpp>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <mutex>
+#include <sstream>
+#include <thread>
+
+#ifdef _WIN32
+# pragma warning ( disable : 4996 )
+#endif
+
+namespace bpt = boost::property_tree;
+namespace {
+// global variables
+std::mutex m;
+uint32_t thread_count;
+} // namespace
+
+namespace xrt_core::usage_metrics {
+
+template <typename met_type, typename find_type>
+static met_type*
+get_metrics(std::vector<met_type>& metrics_vec, const find_type& finder)
+{
+  auto it = std::find_if(metrics_vec.begin(), metrics_vec.end(), 
+                        [finder](met_type& met) 
+                        {
+                          return met.handle == finder;
+                        }
+                        );
+  return it == metrics_vec.end()
+    ? nullptr
+    : &(*it);
+}
+
+struct bo_metrics
+{
+  uint32_t total_count = 0;
+  size_t   total_size_in_bytes = 0;
+  size_t   peak_size_in_bytes = 0;
+  size_t   bytes_synced_to_device = 0;
+  size_t   bytes_synced_from_device = 0;
+};
+
+using tp = std::chrono::time_point<std::chrono::high_resolution_clock>;
+struct kernel_metrics
+{
+  struct timestamp
+  {
+    tp start_time;
+    bool is_valid = false;
+  };
+
+  std::string handle; // kernel name is used as handle for identifying kernel
+  std::vector<uint32_t> cu_index_vec;
+  uint32_t total_runs = 0;
+  std::chrono::microseconds total_time = {};
+  std::unordered_map<const xrt::run_impl*, timestamp> exec_times; // run handle ptr is used for indexing
+  size_t num_args;
+
+  void
+  log_kernel_exec_time(const xrt::run_impl* run_hdl, const tp& tp_now, ert_cmd_state state)
+  {
+    // state ERT_CMD_STATE_NEW indicates kernel start is called
+    if (state == ERT_CMD_STATE_NEW) {
+      // record start everytime because previous run may be finished, timeout, aborted or stopped
+      exec_times[run_hdl].start_time = tp_now;
+      exec_times[run_hdl].is_valid = true;
+    }
+    else {
+      // make start time invalid so we can record for next run, add duration to total time and increment total runs
+      if (exec_times[run_hdl].is_valid && state == ERT_CMD_STATE_COMPLETED) {
+        // valid run increament run
+        total_runs++;
+        total_time += std::chrono::duration_cast<std::chrono::microseconds>(tp_now - exec_times[run_hdl].start_time);
+
+        // invalidate start time, run may be finished, aborted or timed out
+        exec_times[run_hdl].is_valid = false;
+      }
+    }
+  }
+};
+
+struct hw_ctx_metrics
+{
+  const xrt_core::hwctx_handle* handle;  // using hw_ctx handle ptr as unique identifier for logging
+  xrt::uuid xclbin_uuid;
+  struct bo_metrics bos_met;
+  std::vector<struct kernel_metrics> kernel_metrics_vec;
```

Comment:
In C++ the use of `struct` where types are used is not commonly used.   I would recommend just
```suggestion
  std::vector<kernel_metrics> kernel_metrics_vec;
```